OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnProduction.cs

```diff
@@ -0,0 +1,110 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Grants a condition when this actor produces a specific actor.")]
+	public class GrantConditionOnProductionInfo : ITraitInfo
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("The condition to grant")]
+		public readonly string Condition = null;
+
+		[ActorReference]
+		[Desc("The actors to grant condition for. If empty condition will be granted for all actors.")]
+		public readonly HashSet<string> Actors = new HashSet<string>();
+
+		[Desc("How long condition is applies for. Use -1 for infinite.")]
+		public readonly int Duration = -1;
+
+		[Desc("Show a selection bar while condition is applied if it has a duration.")]
+		public readonly bool ShowSelectionBar = true;
+		public readonly Color SelectionBarColor = Color.Magenta;
+
+		public object Create(ActorInitializer init) { return new GrantConditionOnProduction(init.Self, this); }
+	}
+
+	public class GrantConditionOnProduction : INotifyCreated, INotifyProduction, ITick, ISync, ISelectionBar
+	{
+		readonly GrantConditionOnProductionInfo info;
+		ConditionManager manager;
+
+		int token = ConditionManager.InvalidConditionToken;
+		[Sync] int ticks;
+
+		public GrantConditionOnProduction(Actor self, GrantConditionOnProductionInfo info)
+		{
+			this.info = info;
+			ticks = info.Duration;
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			manager = self.TraitOrDefault<ConditionManager>();
+		}
+
+		void GrantCondition(Actor self, string cond)
+		{
+			if (manager == null)
+				return;
+
+			token = manager.GrantCondition(self, cond);
+		}
+
+		void RevokeCondition(Actor self)
+		{
+			if (manager == null)
+				return;
+
+			token = manager.RevokeCondition(self, token);
+		}
+
+		void INotifyProduction.UnitProduced(Actor self, Actor other, CPos exit)
+		{
+			if (info.Actors.Any() && !info.Actors.Select(a => a.ToLowerInvariant()).Contains(other.Info.Name))
+				return;
+
+			if (token == ConditionManager.InvalidConditionToken)
+				GrantCondition(self, info.Condition);
+			else
+				ticks = info.Duration;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (info.Duration >= 0 && token != ConditionManager.InvalidConditionToken)
+			{
+				if (--ticks < 0)
+				{
+					ticks = info.Duration;
+					RevokeCondition(self);
```

Comment:
Likewise here, this should change to `token = manager.RevokeCondition(self, token);`.
An extra check for the condition manager is unnecessary because this is built in to `token != ConditionManager.InvalidConditionToken`.